gpytorch/lazy/lazy_tensor.py

```diff
@@ -775,73 +775,58 @@ def cat_rows(self, cross_mat, new_mat, generate_roots=True, **root_decomp_kwargs
 
-        new_lazy_tensor_1 = CatLazyTensor(old_tsr, cross_mat, dim=-2)
-        new_lazy_tensor_2 = CatLazyTensor(cross_mat.transpose(-1, -2), new_mat, dim=-2)
-        new_lazy_tensor = CatLazyTensor(new_lazy_tensor_1, new_lazy_tensor_2, dim=-1)
+        # form matrix C = [A B; B^T D]
+        upper_row = CatLazyTensor(A, B, dim=-2)
+        lower_row = CatLazyTensor(B.transpose(-1, -2), D, dim=-2)
+        new_lazy_tensor = CatLazyTensor(upper_row, lower_row, dim=-1)
 
         # if the old lazy tensor does not have either a root decomposition or a root inverse decomposition
         # don't create one
-        does_not_have_roots = _is_in_cache_ignore_args(self, "root_decomposition") or _is_in_cache_ignore_args(
-            self, "root_inv_decomposition"
+        does_not_have_roots = any(
+            _is_in_cache_ignore_args(self, key) for key in ("root_inv_decomposition", "root_inv_decomposition")
         )
         if not generate_roots and not does_not_have_roots:
             return new_lazy_tensor
 
-        L = self.root_decomposition(**root_decomp_kwargs).root
-        l_is_triang = isinstance(L, TriangularLazyTensor)
-        L = L.evaluate()
-        L_inverse = self.root_inv_decomposition().root.evaluate()
-
-        m, n = L.shape[-2:]
-
-        if l_is_triang:
-            # The whole thing is triangular, we can just do two triangular solves
-            if isinstance(cross_mat, LazyTensor):
-                cross_mat = cross_mat.evaluate()
-            lower_left = torch.triangular_solve(cross_mat.transpose(-1, -2), L, upper=False).solution.transpose(-2, -1)
-        else:
-            lower_left = cross_mat.matmul(L_inverse)
-
-        schur = new_mat - lower_left.matmul(lower_left.transpose(-2, -1))
-        schur_root = lazify(schur).root_decomposition().root.evaluate()
-
-        # Form new root Z = [L 0; lower_left schur_root]
-        num_fant = schur_root.size(-2)
-        new_root = torch.zeros(*batch_shape, m + num_fant, n + num_fant, device=L.device, dtype=L.dtype)
-        new_root[..., :m, :n] = L
-        new_root[..., m:, : lower_left.shape[-1]] = lower_left
-        new_root[..., m:, n : (n + schur_root.shape[-1])] = schur_root
-
-        # Use pseudo-inverse of Z as new inv root
-        if l_is_triang:
-            # we know L is triangular, so inverting is a triangular solve agaist the identity
-            # we don't need the batch shape here, thanks to broadcasting
-            Eye = torch.eye(new_root.shape[-2], device=new_root.device, dtype=new_root.dtype)
-            new_inv_root = torch.triangular_solve(Eye, new_root, upper=False)[0]
-            new_inv_root = new_inv_root.transpose(-1, -2)
+        E = self.root_decomposition(**root_decomp_kwargs).root
+        m, n = E.shape[-2:]
+        R = self.root_inv_decomposition().root.evaluate()  # this is fast if L is triangular
+        F = B_ @ R
+        G = lazify(D - F.matmul(F.transpose(-2, -1))).root_decomposition().root.evaluate()
```

Comment:
Personally, I find `lower_left` and `schur_root` to be a bit more human comprehensible than `F` and `G` respectively. Along with also doing a better job of pointing out what exactly is going on.

Reply:
The main reason of for changing this was to be consistent with the comments and to dedup some double usage of notation for different things.  I can get back to those var names and add more comments if that is desired.